### PR TITLE
Starts SMS connector

### PR DESCRIPTION
**Pull request by Gideon357**
# Description

I have started building out a SMS connector built on Twilio. I added a dependency on the Twilio helper library. It still needs work, and I haven't tested it. ~~I  wasn't able to find documentation for the web server, so I could have entirely FUBAR\'d it.~~

Fixes: #1490 

## Status
~~READY~~ | **UNDER DEVELOPMENT** | ~~ON HOLD~~

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?
I wrote unittest for this. 

# Todo List

- [ ] Make `Client` async
- [ ] Improve test coverage
- [ ] Improve docs
- [ ] Write a sample application to make sure that it works well

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: opsdroid/connector/sms/__init__.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+_LOGGER = logging.getLogger(__name__)
+CONFIG_SCHEMA = {
+    "name": str,
+    "account_sid": str,
+    "auth_token": str,
+    "phone_number": str,
+    "host": str,
+    "port": int,
+}
```
Comment (jacobtomlinson) [inline opsdroid/connector/sms/__init__.py:20]: I assume some of these are not optional. They should have `voluptuous.Required` around them.

See the Slack connector for [an example](https://github.com/opsdroid/opsdroid/blob/af4253ea542aeec78f86eee426d56abdc937d1ec/opsdroid/connector/slack/__init__.py#L29) of this.

File: opsdroid/connector/sms/__init__.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        self.config = config
+
+    async def connect(self):
+        """Connect to Twilio and setup webhooks"""
+        self.connection = await Client(
+            self.config["account_sid"], self.config["auth_token"]
+        )
+        self.opsdroid.web_server.web_app.router.add_get(
+            f"/connector/{self.name}", self.handle_messages
+        )
```
Comment (jacobtomlinson) [inline opsdroid/connector/sms/__init__.py:39]: This is the correct way to add a get endpoint to the web server. 

Do you have to manually configure Twilio to know what your opsdroid webhook URL is? In a web GUI or something?

File: opsdroid/connector/sms/__init__.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    async def handle_messages(self, request):
+        req_data = await request.json()
+        try:
+            message = Message(req_data["Body"], req_data["From"], None, self,)
+            await self.opsdroid.parse(message)
+        except Exception as e:
```
Comment (jacobtomlinson) [inline opsdroid/connector/sms/__init__.py:46]: This is a pretty broad except, could we constrain this to a simpler set of what we are expecting to go wrong here?

File: opsdroid/connector/sms/__init__.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            self.config["account_sid"], self.config["auth_token"]
+        )
+        self.opsdroid.web_server.web_app.router.add_get(
+            f"/connector/{self.name}", self.handle_messages
+        )
+
+    async def handle_messages(self, request):
+        req_data = await request.json()
+        try:
+            message = Message(req_data["Body"], req_data["From"], None, self,)
```
Comment (jacobtomlinson) [inline opsdroid/connector/sms/__init__.py:44]: It's best to be explicit with kwargs when constructing a Message object.

See the Telegram connector for [an example](https://github.com/opsdroid/opsdroid/blob/af4253ea542aeec78f86eee426d56abdc937d1ec/opsdroid/connector/telegram/__init__.py#L182-L188) of this.

**Comment by FabioRosado** [comment]
This is awesome! I was actually toying with the idea of creating a connector for opsdroid with Twilio so I'm looking forward to see this connector in opsdroid 😄  It would also be awesome to interact with the bot through sms and seen some pretty awesome things done with twilio.

I did a quick look on the twilio python sdk, I believe `Client` isn't async so might not work there 🤔 

Not sure if you know about it, but if you want extra credits to test the bot, you can install twilioquest and do the first level, explore the whole map go to the right and you have a box that contains a code that gives you 50 dollars - should be enough to test things out 😄 👍

File: opsdroid/connector/sms/__init__.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            await self.opsdroid.parse(message)
+        except Exception as e:
+            _LOGGER.error(f"ERROR: {e}")
+
+    @register_event(Message)
+    async def send_message(self, message):
+        try:
+            self.connection.messages.create(
```
Comment (jacobtomlinson) [inline opsdroid/connector/sms/__init__.py:52]: This looks like a sync function call. If there is no async client for twilio then this should be run in an executor.

File: opsdroid/connector/sms/__init__.py (hunk 2 above)
Comment (Gideon357) [reply opsdroid/connector/sms/__init__.py:39]: Yes, I couldn't find a way to do it via the REST api.

File: opsdroid/connector/sms/__init__.py (hunk 1 above)
Comment (Gideon357) [reply opsdroid/connector/sms/__init__.py]: @jacobtomlinson thanks for reviewing this! I have resolved this

File: opsdroid/connector/sms/__init__.py (hunk 4 above)
Comment (Gideon357) [reply opsdroid/connector/sms/__init__.py]: Finished

**Comment by FabioRosado** [comment]
Hello @Gideon357 the stale bot closed this PR, how are you getting on with the connector? Do you need any help from us?

**Comment by Gideon357** [comment]
Hi @FabioRosado sorry about that. Some stuff came up and I haven't been able to work on this. I might have some time today, so I will get back to you. Thanks for all your help! Once again, I am really sorry about that.

**Comment by Gideon357** [comment]
@FabioRosado I was able to add documentation, but I am not sure if / when I will be able to complete this. I have added a checklist here, and will update issue #1490 so if anyone wants to finish this they can. Sorry about that! I look forward to continuing to work on @opsdroid.

**Comment by FabioRosado** [comment]
Hello @Gideon357  thank you for the message and apologies for not replying to you earlier. It's all good, we know that things come up, I just wanted to check up on you and see if you need any help.

I'll leave the PR as it is until you are able to pick it up, if you are unable I might try and pick it up from where you left it 😄 👍

**Comment by danielballan** [comment]
I am interested in picking this up. My use case is a hobby one, not a day-job one, so I would be proceeding at the rate of an evening/weekend project. Is there anything I should know before attempting to carry on from where the author left off?

**Comment by FabioRosado** [comment]
Hello @danielballan  thank you for showing interest in helping the project with this. This PR was a great start to implement the SMS connector. 

I'd advise you to check Jacob's comments on Gideon code, there were some tests failing including linting if memory doesn't fail. We have changed quiet a few things to opsdroid so I'm not sure how easy it will be to pick up this PR, you could give it a try though 👍 

If you have any issue or need any help with anything you can reach us on our [matrix channel](https://riot.im/app/#/room/#opsdroid-general:matrix.org) or you can also submit a WIP PR and we can talk there 👍

**Comment by danielballan** [comment]
Great, that’s the context I was looking for. I’ll take a look and weigh whether to rebase this or just use it as a reference for a fresh start.

**Comment by jacobtomlinson** [comment]
> My use case is a hobby one, not a day-job one, so I would be proceeding at the rate of an evening/weekend project

That's pretty much everyone here :)

**Pull request closed without merging**